### PR TITLE
Multi-worker support for Resnet.

### PR DESCRIPTION
**Pull request by dubey**
This change adds support for running on multiple workers using `CollectiveAllReduceStrategy`.  It adds 2 flags to `resnet_run_loop.py`: `worker_hosts` and `task_index`.  These flags provide information to each worker about peers in the multi-worker setup.

After this change, when the official resnet model runs with `tf.contrib.distribute.CollectiveAllReduceStrategy,` it uses `tf.estimator.train_and_evaluate` rather than separate `train` and `evaluate` calls.

### CODE REVIEW

File: official/resnet/resnet_run_loop.py [hunk 1]
```diff
@@ -563,24 +584,41 @@ def input_fn_eval():
...
     # eval (which is generally unimportant in those circumstances) to terminate.
     # Note that eval will run for max_train_steps each loop, regardless of the
     # global_step count.
-    eval_results = classifier.evaluate(input_fn=input_fn_eval,
-                                       steps=flags_obj.max_train_steps)
+    if use_train_and_evaluate:
+      eval_spec = tf.estimator.EvalSpec(input_fn=input_fn_eval,
+                                        steps=flags_obj.max_train_steps)
+      tf.compat.v1.logging.info('Starting to train and evaluate.')
+      eval_results, _ = tf.estimator.train_and_evaluate(classifier, train_spec,
```
Comment (yuefengz) [inline official/resnet/resnet_run_loop.py]: I am not sure whether it is the best practice to call `train_and_evaluate` in a loop. Usually we make an input_fn for all epochs and training finishes when `train_and_evaluate` returns. Also `train_and_evaluate` would run evaluation on evaluator continuously until it reaches  `max_train_steps`.

File: official/resnet/resnet_run_loop.py [hunk 2]
```diff
@@ -563,24 +584,41 @@ def input_fn_eval():
     schedule = [flags_obj.epochs_between_evals for _ in range(int(n_loops))]
     schedule[-1] = flags_obj.train_epochs - sum(schedule[:-1])  # over counting.
 
+  use_train_and_evaluate = isinstance(
```
[the remark this answers is not in the document]
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Sure, I'll make that change in the next commit.

File: official/resnet/resnet_run_loop.py (hunk 1 above)
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Can you elaborate a bit more? Does this mean when using `train_and_evaluate` there is no loop? But then how do we specify number of epochs?

File: official/resnet/resnet_run_loop.py [hunk 3]
```diff
@@ -468,6 +469,25 @@ def resnet_main(
   if flags_obj.tf_gpu_thread_mode:
     override_flags_and_set_envars_for_gpu_thread_pool(flags_obj)
 
+  # Configures cluster spec for distribution strategy.
+  tf_config = json.loads(os.environ.get('TF_CONFIG', '{}'))
```
[the remark this answers is not in the document]
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Will do in next commit.

File: official/resnet/resnet_run_loop.py [hunk 4]
```diff
@@ -645,6 +683,15 @@ def define_resnet_flags(resnet_size_choices=None):
           'the expense of image resize/cropping being done as part of model '
           'inference. Note, this flag only applies to ImageNet and cannot '
           'be used for CIFAR.'))
+  flags.DEFINE_string(
```
[the remark this answers is not in the document]
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: ack

File: official/resnet/resnet_run_loop.py (hunk 1 above)
Comment (yuefengz) [reply official/resnet/resnet_run_loop.py]: We specify the number of epochs in the `input_fn` or specify the number of steps in `train_spec`. @karmel can correct me if I understand it wrong.

File: official/utils/misc/distribution_utils.py [hunk 5]
```diff
@@ -56,6 +58,10 @@ def get_distribution_strategy(distribution_strategy="default",
                        "cannot be set to 'off'.".format(num_gpus))
     return None
 
+  if distribution_strategy == "collective" or num_workers > 1:
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: is 1 remote worker a valid use case? which strategy do we want users to use in that case? 

also, will it be confusing if the user explicitly says "collective" with 1 worker, but we still fallback to mirrored strategy? At the very least we should say if this is the case in documentation.

File: official/resnet/resnet_run_loop.py [hunk 6]
```diff
@@ -645,6 +683,15 @@ def define_resnet_flags(resnet_size_choices=None):
...
+      name='worker_hosts', default=None,
```
Comment (guptapriya) [inline official/resnet/resnet_run_loop.py]: I guess the expectation is that the user would have started this program on all these hosts? If yes, should we clarify this?

File: official/resnet/resnet_run_loop.py [hunk 7]
```diff
@@ -478,6 +498,7 @@ def resnet_main(
   distribution_strategy = distribution_utils.get_distribution_strategy(
       distribution_strategy=flags_obj.distribution_strategy,
       num_gpus=flags_core.get_num_gpus(flags_obj),
+      num_workers=num_workers,
       all_reduce_alg=flags_obj.all_reduce_alg)
 
   # Creates a `RunConfig` that checkpoints every 24 hours which essentially
```
Comment (guptapriya) [inline official/resnet/resnet_run_loop.py]: Unrelated to your CL but i just realized we do not distribute evaluation in this case. Should we? Or is it a bad idea because we throw away any partial batch so eval may not be always complete based on the batch size..

File: official/resnet/resnet_run_loop.py [hunk 8]
```diff
@@ -645,6 +683,15 @@ def define_resnet_flags(resnet_size_choices=None):
           'the expense of image resize/cropping being done as part of model '
           'inference. Note, this flag only applies to ImageNet and cannot '
           'be used for CIFAR.'))
+  flags.DEFINE_string(
+      name='worker_hosts', default=None,
+      help=flags_core.help_wrap(
+          'Comma-separated list of worker ip:port pairs for running '
+          'multi-worker models with DistributionStrategy.'))
+  flags.DEFINE_integer(
```
Comment (guptapriya) [inline official/resnet/resnet_run_loop.py]: Are you planning to update the README in this directory to guide users how to use multi worker training? (Can be a separate PR but would be good to update)

File: official/utils/misc/distribution_utils.py (hunk 5 above)
Comment (dubey) [reply official/utils/misc/distribution_utils.py]: The current behavior is that we use `CollectiveAllReduceStrategy` either when user explicitly specifies `distribution_strategy=collective` or there is more than 1 worker.  So, if they specify `collective` with 1 worker it should use `CollectiveAllReduceStrategy` and not `MirroredStrategy`.

File: official/resnet/resnet_run_loop.py (hunk 6 above)
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Yes, if they are launching the program directly they need to issue similar invocations at all workers.

File: official/resnet/resnet_run_loop.py (hunk 8 above)
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Thanks for that suggestion, will do

File: official/resnet/resnet_run_loop.py (hunk 6 above)
Comment (guptapriya) [reply official/resnet/resnet_run_loop.py]: Should we clarify this in the docstring?

File: official/utils/misc/distribution_utils.py (hunk 5 above)
Comment (guptapriya) [reply official/utils/misc/distribution_utils.py]: oh my bad. I think I must've read that "or" as an "and". What you have is fine then.

File: official/utils/misc/distribution_utils.py [hunk 9]
```diff
@@ -56,6 +58,10 @@ def get_distribution_strategy(distribution_strategy="default",
...
+    return tf.contrib.distribute.CollectiveAllReduceStrategy(
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: If Yuefeng submits his change to export the `CollectiveAllReduceStrategy` (and ParameterServerStrategy) in core TF soon, then perhaps this can be changed to use that and be 2.0 compatible?

File: official/resnet/resnet_run_loop.py (hunk 1 above)
Comment (karmel) [reply official/resnet/resnet_run_loop.py]: My understanding matches @yuefengz 's -- train and eval runs the loop for you.

File: official/resnet/resnet_run_loop.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Done.

File: official/resnet/resnet_run_loop.py (hunk 1 above)
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Thanks, done.

File: official/resnet/resnet_run_loop.py (hunk 6 above)
Comment (dubey) [reply official/resnet/resnet_run_loop.py]: Done.

File: official/utils/misc/distribution_utils.py (hunk 9 above)
Comment (dubey) [reply official/utils/misc/distribution_utils.py]: Sure, I will update after Yuefeng's change.

**Comment by dubey** [comment]
I think I've resolved all blockers for this PR.  @yuefengz PTAL.

File: official/utils/misc/distribution_utils.py [hunk 10]
```diff
@@ -56,6 +60,10 @@ def get_distribution_strategy(distribution_strategy="default",
                        "cannot be set to 'off'.".format(num_gpus))
     return None
 
+  if distribution_strategy == "collective" or num_workers > 1:
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: @yuefengz @dubey Since the name of the strategy as exposed has now changed (MultiWorkerMirroredStrategy), i wonder if we should rename this flag too?

Comment (dubey) [reply official/utils/misc/distribution_utils.py]: done.  I'll change the strategy object from contrib.distribute.collectiveallreducestrategy to distribute.multiworkermirroredstrategy in a follow-on PR.
